### ISSUE TITLE
No unit/functional/e2e tests

### ISSUE THREAD

**Issue by thedrow** [tumblr/jetpants#23]
I know you guys are using this in production but it's really hard to contribute from the outside without knowing if something is broken.
Unless I'm missing something this project has no tests at all.
This project needs coverage.
In order to start doing that we should enable travis and coveralls to track when and if tests fail.

**Comment by KirilAngov** [comment]
This is our internal goal as well. We are looking seriously into at least starting the unit tests so that other people can contribute as well. I realize this is way overdue and something we need to work on.

**Comment by thedrow** [comment]
I think that BDD specs will be way more valuable.
If somone (like me) wanted to add Postgres support or port the tool to another language (possibly Go is a good fit) it would just mean I have to copy the specs and ensure all tests pass.
Unit tests are valuable but they will serve another purpose for this project. They will ensure no regression in the algorithms presented here may occur.
The real problem is that adding new features from the outside (and I do believe that this applies from within Tumblr as well, correct me if I'm wrong) is that there's is no full specification of what this product does that can be verified.

**Comment by eelias** [comment]
I'd agree with Kiril that unit tests would be valuable, and are practical in most cases here. They were mainly only omitted originally due to time and resource constraints, especially since `jetpants console` makes ad-hoc testing of changes reasonably straightfoward.

However, I'm far less convinced that functional testing or end-to-end/integration testing is practical enough to pursue here. Although achievable in theory, this is enormously complex and time-consuming for database automation, and that's not even accounting for environmental complexities of making the testing generic enough for anyone to run. Many operations in Jetpants deal with performing complex multi-step automation on 2 to 6 distinct database servers at a time, and testing that effectively requires some pretty deep custom integration into a company's provisioning, service discovery, and hardware management systems -- not to mention having dependencies on specific Linux distributions, MySQL forks and versions, and so forth.

Even aside from that, database automation is generally quite difficult to test effectively -- databases are stateful, counter-productive to mock / test via DI, and have an absolutely massive array of possible failure domains. I can say authoritatively that most complex database automation has, at best, partial test coverage coupled with a high rate of false positives / flaky tests -- and most database automation isn't even open source, nor made generic enough to even run in another company's environment, let alone have test coverage for one.

TL;DR -- yes the current lack of tests should be rectified, and the difficulty in making outside contributions is a 100% valid complaint, but I don't believe that BDD specs are the answer. The vast majority of open source software simply does not have a full, verifiable specification of what the project does. Likewise, with the vast majority of open source software, I wouldn't expect ease of porting the software to another database or language to be high on the priority list of the current maintainers (or their employers) unless doing so is already aligned with their tech stack or business requirements.

**Comment by thedrow** [comment]
You have raised multiple important issues I'd like to address:
1) How to support multiple databases/versions/operating systems?
We can test multiple databases/operating systems combinations using [test kitchen](http://kitchen.ci/).
I can contribute an initial test suite that provisions all the databases we support. Once Travis provides KVM support (which is already underway) you'll be able to run all of those in the CI.
Until that happens we can run the test suite locally.
2) Flanky tests and false positives
I'm not saying we can catch all errors and problems but at least we'll be able to verify the "happy path" for some of the features.
If I insert 10 rows into a single database and reshard it using N instances I expect to have 10 / N rows in each database provided that N is even. Are these kind of tests likely to be flanky?
An even more relaxed requirement would be to not have all 10 rows in the same database instance if the previous expactation can be unreliable. Would you consider that a false positive?
3) Lack of deep integration into other companies infrastructure
Does this mean that this tool is specific to Tumblr's usecase? I'd certainly hope not.
How does this tool relate to service discovery? As far as I can see this tool does not provide any integration with service discovery and does not auomatically know where you're database instances are.
Can you explain how does your provisioning, service discovery and hardware management affect testing of this software?
3) Most OSS software project don't have verifiable spec
Most OSS projects don't need one. However porting Hub to Go was mainly successful because of those specs.
If someone outside Tumblr would like to port this to another language then specifications would be very useful.
4) Priorities of maintainers
I do not expect the maintainers to contribute PostgreSQL support if it doesn't fit their stack and I do not expect them to merge a port to another language that they are not familiar with but to enable those options for other people in the spirit of open source.
The fact that I need this tool to work with PostgreSQL does not obligate the maintainers to implement such a feature. It does add a bit of maintainece burden of merging pull requests that are related to PostgreSQL which may be unacceptable if automated testing is not possible since merging code requires us to know that a feature works or a bug fix actually fixes a bug.
The maintainence burden might be unacceptable anyway and in which case, I'd like to know so that I'll look elsewhere for a project that will support PostgreSQL.

**Comment by eelias** [comment]
> Until that happens we can run the test suite locally.

This would involve running a minimum of a half-dozen different virtualbox instances on the same local machine -- it's pretty heavy. Could possibly use docker containers instead, but the setup is nontrivial, especially if testing any real data size. Even more complex when accounting for interacting with service discovery, provisioning, etc -- systems which tend to differ at every company.

> If I insert 10 rows into a single database and reshard it using N instances I expect to have 10 / N
> rows in each database provided that N is even. Are these kind of tests likely to be flanky?

Shard splitting (via Jetpants' technique) is a multi-step operation involving provisioning servers, cloning replicas, exporting data, dropping data, re-importing data, cloning more replicas, making repeated service discovery changes, manipulating replication, and then cleaning up superfluous replicated data. It is not a trivial operation, and therefore neither is testing it.

I have seen (and, to a lesser extent, worked on) other MySQL shard-splitting techniques, which involve different steps but are generally equally complex.

As I mentioned, the problem is the range of possible failure vectors. To clarify, with a single client doing a single write to a single database, failures are individually quite rare, but the number of *possible* failures is quite large: the server may run out of disk space, the network may have issues, the client may be overloaded, the client may be unable to talk to service discovery to find out where the server is, etc. Consider also the complication of "soft failures", where the client just gets a timeout and has no way of knowing if their operation succeeded or failed.

When talking about automation across a half-dozen hosts and interacting with other external services, the range of possible failure scenarios grows exponentially. Combine this with running tests continuously, and you'll get an annoying rate of flaky tests, even if throwing a tremendous amount of manpower at the problem. I can say this authoritatively, having spent a good chunk of the past 4+ years working on database automation for two of the world's largest social networks. 

> An even more relaxed requirement would be to not have all 10 rows in the same database 
> instance if the previous expactation can be unreliable. Would you consider that a false positive?

Database automation is (or at least, certainly should be) deterministic. The possible issue is failing part-way through a many-step non-idempotent process, not problems with randomly incorrect behavior.

> Does this mean that this tool is specific to Tumblr's usecase? 

No. Please read the documentation, specifically the docs on plugins, requirements, FAQ, and the main readme.

> How does this tool relate to service discovery?

In brief, a tool that manipulates shards needs both read and write access to the mapping of shards to database instances. This is documented, and also mentioned in numerous slide decks. 

> If someone outside Tumblr would like to port this to another language then specifications 
> would be very useful.

IMO, porting it to another language as-is doesn't make much sense. It's very much geared towards Ruby, and taking advantage of unique features of the language. The organization of the library would be entirely different in idiomatic Go.

That's not to say it wouldn't be awesome to have a Go-based open source database automation library/toolkit. Just saying it would be structured entirely different than Jetpants. It could certainly implement some of the same operations, but the structural differences mean that BDD specs are not necessarily any more useful than the existing documentation and slide decks.

Let's take a step back: what specifically are you trying to achieve, and why? Knowing that will help guide solutions.

**Comment by thedrow** [comment]
As it currently stands using a vanilla PostgreSQL setup in a medium/high scale environment isn't easy.
You have alternatives like CitrusDB and PostgresXL but they are not available in RDS which means that they require us to maintain the database machines ourselves. Most companies are not in a stage where they need to do that yet.
Unfortunately, since "boring" technologies are preferred in my company and probably a lot of other companies as well other databases that will fit the kind of workload that requires a little bit of sharding (3-6 nodes) or read/write replicas such as Cassandra are not an option.
The only viable solution in these environment is pgpool but it only supports [one sharding scheme](https://gist.github.com/afair/3803895#parallel-query-for-sharding-and-partitioning).

I think the community needs a go-to solution for sharding and replication that doesn't involve installing extensions in PostgreSQL since that won't be supported on RDS.
In order to achieve that I think that adding PostgreSQL support to this project will help everyone who's using PostgreSQL in an environment that requires high availability and/or sharding.

Even if you do use extensions such as [pg_shard](https://github.com/citusdata/pg_shard) on your own machines running somewhere you need a toolkit that abstracts the different commands that pg_shard exposes and orchestration for the kind of complex operations that jetpants provides you with.

So the first goal is to add PostgreSQL support that will match as closely as possible to what jetpants provides for MySQL.
In order to do that I'd need to ensure that I haven't broken anything and that the code I wrote works and does what I intended to do.
Unit testing alone probably won't provide me the sense of confidence I need.

The ultimate goal is to have a generic database automation tool that supports multiple service discovery tools and configuration management tools and that automatically configures database proxies such as pgpool or pgbouncer. Another option would be to write a database proxy from scratch that doesn't need to be restarted when you add a new node through service discovery and is already aware of every sharding scheme we want to support.

This is why Go comes to mind. It has a library called [libkv](https://github.com/docker/libkv) that can communicate with all common service discovery tools, it has a concurrency model that is very applicable to what we're already doing and to the features that I'd like to see in the perfect database automation/administration tool.
That tool should of course keep supporting MySQL and it's variants.
I recognize that Tumblr might or might not want to maintain PostgreSQL support for jetpants so porting the project might be a more viable option. I'll be happy to keep the code in Ruby if some of the features I described make sense to you guys.
